### ISSUE TITLE
attachment settings create new folder when directory space limit is reached

### ISSUE THREAD

**Issue by illori** [SimpleMachines/SMF#597]
for ease of testing, set the method for management of the attachment directories to auto when either directory space limit is reached 

set max number of files per directory to a small number lets say 5

upload a total of 5 attachments to posts

verify how many attachments end up in the folders.

you will see a total of 5 files but that includes the index.php and the .htaccess IMO it should exclude those 2 files

also using this option it does not use the default attachments folder it remains empty

you can also verify the number of files smf sees in the folders by going to the attachment directories page in the attachments and avatars page in admin.

**Comment by jdarwood007** [comment]
Looking at the code it doesn't check the directory,  It is relying on the database to return the proper count.  Is it actually creating the new directories for those files or are you just seeing that in the admin interface?

**Comment by illori** [comment]
it is creating new directories

**Comment by Oldiesmann** [comment]
So if you upload 5 files, it's putting 3 in one folder then creating a new folder for the other 2?

**Comment by illori** [comment]
yes exactly

**Comment by Oldiesmann** [comment]
After some digging, it seems the issue is due to this code (lines 590 and 591 of Subs-Attachments.php):

``` php
        if (!empty($modSettings['attachmentDirFileLimit']) && $context['dir_files'] + 2 > $modSettings['attachmentDirFileLimit']
            || (!empty($modSettings['attachmentDirSizeLimit']) && $context['dir_size'] > $modSettings['attachmentDirSizeLimit'] * 1024))
```

Since the setting is "maximum number of files per directory", SMF is sticking to that - if you tell it not to put more than 5 files per directory, it does exactly that - no more than 5 files per directory, counting the 2 that are added by SMF to prevent people from accessing the folder directly.

Whether this is actually a bug or not is a matter of opinion though as whoever added these settings didn't clarify what they were intending to do with it - the helptext for that section doesn't cover it. It also doesn't really cause any problems other than creating a few extra directories. Even if you have it set to 100 files per folder and 10,000 attachments, that's only 3 extra folders needed to store the attachments.

If we're going to fix this, we really need to clarify that setting (though at this point I'd also like to change how the helptext is handled there - it makes no sense to have one that covers every setting in that section instead of one for each individual setting).

**Comment by Arantor** [comment]
Wait, why are we adding 2 to $context['dir_files']?

We're using the DB count of what's in the folder, not the physical file count (and if we _were_ using the physical file count, then we should be subtracting, not adding, 2 anyway... or at least adding 2 on the correct side of the comparison)...

IMHO, we should never be counting the .htaccess or index.php files, they're not relevant to the number of attachments and cause unexpected behaviour as described.

The helptext has been split up, don't know if we really want to change them much more.

**Comment by Oldiesmann** [comment]
Is this something we want to fix or not? SMF is sticking to the "max files in directory" setting - if you tell it to allow 5 files, it will allow exactly 5 files, counting the .htaccess and index.php files that it also adds. I'm thinking it would be best to make that setting apply to how many actual attachment files are in the directory, rather than counting both index.php and .htaccess in there as well.

**Comment by Arantor** [comment]
It should not count .htaccess or index.php in that as outlined before :tongue:

**Comment by Oldiesmann** [comment]
Ok. Will fix this as well.

**Comment by Arantor** [comment]
Please can someone retest this before I close it?

**Comment by Xarcell** [comment]
Here's what did to be sure that I'm trying to duplicate the issue.

1.| Downloaded the latest repo & fresh installed.

2.| Admin > Forum > Attachments Settings > Maximum number of files per directory > Set To 5 (instead of the default 1000)

3.| I create 1 topic, with 4 replies, each with their own attachment.

4.| The attachments directory has 5 attachments, and an index & htaccess file.

5.| I created another reply to the same topic with another attachment.

6.| The attachment directory has 6 attachments, 8 files total.

7.| I created another topic with 4 attachments. Now the attachment directory has 10 attachments, 12 files total.

Attachment directory has exceeded 5 files, and no new directory is created. On the other hand, it's not counting the index & htaccess files as part of the total count.
